feat: added prefix to look for containerid

abhee11 opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

-https://github.com/open-telemetry/opentelemetry-js-contrib/issues/2339

## Short description of the changes

- adding some prefix check - if they exist to truncate them from the container id string -

abhee11 commented:
@blumamir @rauno56 @dyladan Can I get a review or you can assign it to a relevant person

blumamir commented on detectors/node/opentelemetry-resource-detector-container/src/detectors/ContainerDetector.ts:
```diff
@@ -94,22 +81,70 @@ export class ContainerDetector implements Detector {
+        }
+        // Remove anything after the first period
+        if (lastSection.includes('.')) {
+          lastSection = lastSection.split('.')[0];
```
the original code is using the last dot:

```js
        let startIdx = lastSection.lastIndexOf('-');
        let endIdx = lastSection.lastIndexOf('.');

        startIdx = startIdx === -1 ? 0 : startIdx + 1;
        if (endIdx === -1) {
          endIdx = lastSection.length;
        }
        if (startIdx > endIdx) {
          continue;
        }
```

Can you elaborate about the motivation for this change, perhaps also include some examples of what this`section` might be formatted like to give more context for the review and future maintainers?

abhee11 replied:
So the motivation for this change is as follows:

1. Add more cases to be handled ( more prefixes ) - previous case only handled a specific case
2. Adding more filter - knowing that v1 id is hexadecimal in nature - adding that check 
3. The new logic is essentially doing very similar but with more checks
4. Added a lot of tests and example in utils.test.ts

blumamir commented on detectors/node/opentelemetry-resource-detector-container/src/detectors/ContainerDetector.ts:
```diff
@@ -51,35 +56,17 @@ export class ContainerDetector implements Detector {
-        // https://github.com/containerd/containerd/blob/release/1.5/pkg/cri/server/helpers_linux.go#L64
-        return lastSection.substring(colonIdx + 1);
-      } else {
-        let startIdx = lastSection.lastIndexOf('-');
```
was the search for `-` intended to remove the prefixes (`docker-` etc)?
can the new code now fail to remove such prefixes for something which is not in the predefined list of know prefixes?

abhee11 replied:
we are taking care of more prefixes :   
  readonly CRIO = 'crio-';
  readonly CRI_CONTAINERD = 'cri-containerd-';
  readonly DOCKER = 'docker-';

earlier code only took care of docker - this change handles additional cases.

blumamir commented on detectors/node/opentelemetry-resource-detector-container/src/detectors/ContainerDetector.ts:
```diff
@@ -94,22 +81,70 @@ export class ContainerDetector implements Detector {
+        // Handle containerd v1.5.0+ format with systemd cgroup driver
+        const colonIndex = lastSection.lastIndexOf(':');
+        if (colonIndex !== -1) {
+          lastSection = lastSection.substring(colonIndex + 1);
```
previous code behavior was to return in this case, and avoid any other manipulations on the `lastSection`.

Since I don't have any prior knowledge on how containers are named and formatted, it's difficult to review this change.
Can you please describe the reasoning behind these changes, preferably with examples

blumamir commented on detectors/node/opentelemetry-resource-detector-container/src/detectors/ContainerDetector.ts:
```diff
@@ -94,22 +81,70 @@ export class ContainerDetector implements Detector {
-    return containerIdStr || '';
+
+    const strArray = str?.split('/') ?? [];
+    for (let i = 0; i < strArray?.length; i++) {
```
if `strArray` is `undefined`, then `strArray?.length` is also `undefined`. not sure what will come up from `<` a number to `undefined`, but I don't think we should keep it.
Actually, since strArray is nullish coalesced, it cannot be `undefined` so using optional chaining is not needed here.

you can choose to refactor this section to simplify the flow in case `str` is empty, and then return instead of iterating an empty list.

blumamir commented on detectors/node/opentelemetry-resource-detector-container/src/detectors/ContainerDetector.ts:
```diff
@@ -94,22 +81,70 @@ export class ContainerDetector implements Detector {
+    for (let i = 0; i < strArray?.length; i++) {
+      if (
+        strArray[i] === this.MARKING_PREFIX &&
+        strArray[i + 1]?.length === this.CONTAINER_ID_LENGTH
```
for the last iteration, `i+1` is always out of range and will be undefined.
We can change the loop so it ends at `strArray.length - 1` instead of accessing index out of range.

blumamir commented on detectors/node/opentelemetry-resource-detector-container/test/ContainerDetector.test.ts:
```diff
@@ -146,5 +146,75 @@ describe('ContainerDetector', () => {
+          '11:devices:/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod5c5979ec_6b2b_11e9_a923_42010a800002.slice/crio-1234567890abcdef.scope';
+        const expected = '1234567890abcdef';
+        assert.strictEqual(
+          containerDetector['extractContainerIdFromLine'](line),
```
Accessing the private / protected function this way feels wrong. 
WDYT about extracting the function to `utils.ts` instead of it being class member function. then we can test it directly in an idiomatic way.

abhee11 replied:
it does feel wrong. I shall move the helper methods as per suggestion.

blumamir reviewed (commented):
I have very little knowledge about how containers are stored and formatted in files. added few general suggestions / concerns.

The new tests seems not to cover some of the flows introduced by this PR. Please consider adding more tests.

abhee11 commented:
@blumamir Thank you for the review - I have added some comments and added more test cases in utils.test.ts This PR 

1. Covers more prefixes  ( previous was only docker ) this has more coverage.
2. Added additional hexadecimal checks. 
3. Added suffix check and improved the start and end consideration of what might be containerID 
4. Added more tests to cover new functionality .

blumamir reviewed (approved):
Thank you for working on this and addressing all the comments.
LGTM
